src/System.Sorting/src/TimSort.cs

```diff
@@ -0,0 +1,680 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Sorting;
+using System.Text;
+
+// This implementation was referenced from Java opensource, which has been found here:
+// http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8-b132/java/util/TimSort.java
+// Other Open Source references used C++:
+//https://github.com/gfx/cpp-TimSort/blob/master/timsort.hpp
+//Reference: https://en.wikipedia.org/wiki/Timsort
+//ref:http://svn.python.org/projects/python/trunk/Objects/listsort.txt - Original explanation from Tim
+//Timsort is a hybrid stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of partially sorted real-world data
+//Timsort is not a standalone algorithm but a hybrid, an efficient combination of a number of other algorithms
+//The algorithm finds subsets of the data that are already ordered, and uses that knowledge to sort the remainder more efficiently. This is done by merging an identified subset, called a run, with existing runs until certain criteria are fulfilled
+namespace System.Sorting
+{
+    //utility class for Timsort
+ internal static class TimSortUtilities
+    {
+        //?Jave uses 32, Timsort Python originally uses 64. this may need to be revisited by testing new numbers for C# 
+        //several magic numbers used for efficiency
+        //A natural run is a sub-array that is already ordered. Timsort chooses a sorting technique depending on the length of the run. For example, if the run length is smaller than a certain value, insertion sort is used. Thus Timsort is an adaptive sort
+        //If the entire array is less than this length, no merges will be performed, instead Binary Insertion sort is used directly.
+        internal const int MinMerge = 32;
+
+        //Most of the merge occurs in what is called "one pair at a time" mode, 
+        //where respective elements of both runs are compared. When the algorithm merges left-to-right, 
+        //the smaller of the two is brought to a merge area. A count of the number of times the final element appears in a given run is recorded. When this value reaches a certain threshold, MIN_GALLOP, the merge switches to "galloping mode".
+        internal const int MinGallop = 7;
+
+        //C uses 85, Java uses a mixed mode of arriving at most efficient number for stackLength. This varies depending on 
+        //MinMerge number above. we may need to readjust these nos, in c# based on code performance.
+        internal const int StackLength = 40;
+
+        //internal T[] mergeBufferArray=new T[256];
```

Comment:
There are a bunch of places where you have commented out code like this.  Can you remove them?